packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,306 @@
+  // In that case, onWorkScheduled may be called more than once before onWorkStopped.
+  // Work is scheduled by a "thread" which is identified by a unique ID.
+  onWorkScheduled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of scheduled work has been canceled.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkCanceled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has started for a set of interactions.
+  // When this work is complete, onWorkStopped will be called.
+  // Work is not always completed synchronously; yielding may occur in between.
+  // A batch of async/yieldy work may also be re-started before completing.
+  // In that case, onWorkStarted may be called more than once before onWorkStopped.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStarted: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has completed for a set of interactions.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStopped: (interactions: Set<Interaction>, threadID: number) => void,
+};
+
+export type InteractionsRef = {
+  current: Set<Interaction>,
+};
+
+export type SubscriberRef = {
+  current: Subscriber | null,
+};
+
+const DEFAULT_THREAD_ID = 0;
+
+// Counters used to generate unique IDs.
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef = (null: any);
+
+// Listener(s) to notify when interactions begin and end.
+// Note that subscribers are only supported when enableInteractionTrackingObserver is enabled.
+let subscriberRef: SubscriberRef = (null: any);
+
+if (enableInteractionTracking) {
+  interactionsRef = {
+    current: new Set(),
+  };
+  if (enableInteractionTrackingObserver) {
+    subscriberRef = {
+      current: null,
+    };
+  }
+}
+
+// These values are exported for libraries with advanced use cases (i.e. React).
+// They should not typically be accessed directly.
+export {interactionsRef as __interactionsRef, subscriberRef as __subscriberRef};
+
+export function clear(callback: Function): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const prevInteractions = interactionsRef.current;
+  interactionsRef.current = new Set();
+
+  try {
+    return callback();
+  } finally {
+    interactionsRef.current = prevInteractions;
+  }
+}
+
+export function getCurrent(): Set<Interaction> | null {
+  if (!enableInteractionTracking) {
+    return null;
+  } else {
+    return interactionsRef.current;
+  }
+}
+
+export function getThreadID(): number {
+  return ++threadIDCounter;
+}
+
+export function subscribe(subscriber: Subscriber): void {
```

Comment:
I kind of agree with @acdlite here.

We could have a separate module. `interaction-tracker/observer` that attaches itself to `__subscriberRef` directly but provides multi-listener support.

That could be the safe API. That way the minimal module remains even tinier by excluding the overhead of exposing these methods and this error message.

It's never really safe to listen to a single listener protocol directly regardless. The safe form is the multi-listener form.

Reply:
Ok. I'll pull the `subscribe`/`unsubscribe` methods for now and re-add them in a follow up.